File: app/src/main/java/hr/kacan/trznica/repository/LoginRepository.java
package hr.kacan.trznica.repository;

import androidx.lifecycle.MutableLiveData;

import hr.kacan.trznica.data.Result;
import hr.kacan.trznica.models.Korisnik;
import hr.kacan.trznica.data.LoginDataSource;


/**
 * Class that requests authentication and user information from the remote data source and
 * maintains an in-memory cache of login status and user credentials information.
 */
public class LoginRepository {

    private static volatile LoginRepository instance;

    private LoginDataSource dataSource;

    private Korisnik user = null;

    // private constructor : singleton access
    private LoginRepository(LoginDataSource dataSource) {
        this.dataSource = dataSource;

    }

    public static LoginRepository getInstance(LoginDataSource dataSource) {
        if (instance == null) {
            instance = new LoginRepository(dataSource);
        }
        return instance;
    }

    public boolean isLoggedIn() {
        return user != null;
    }

    public void logout() {
        user = null;
        dataSource.logout();
    }

    private void setLoggedInUser(Korisnik user) {
        this.user = user;

    }

    public MutableLiveData<Result<Korisnik>> login(String username, String password) {
        // handle login
        MutableLiveData<Result<Korisnik>> result = dataSource.login(username, password);

        if (result.getValue() instanceof Result.Success) {
            setLoggedInUser(((Result.Success<Korisnik>) result.getValue()).getData());
        }
        return result;
    }
}

File: app/src/main/java/hr/kacan/trznica/repository/LoginRepository.kt
package hr.kacan.trznica.repository

import androidx.lifecycle.MutableLiveData
import hr.kacan.trznica.models.Korisnik
import hr.kacan.trznica.view.login.LoginDataSource
import hr.kacan.trznica.view.login.Result

/**
 * Class that requests authentication and user information from the remote data source and
 * maintains an in-memory cache of login status and user credentials information.
 */
class LoginRepository
private constructor(private val dataSource: LoginDataSource) {
    private var user: Korisnik? = null
    fun isLoggedIn(): Boolean {
        return user != null
    }

    fun logout() {
        user = null
        dataSource.logout()
    }

    private fun setLoggedInUser(user: Korisnik?) {
        this.user = user
    }

    fun login(username: String, password: String): MutableLiveData<Result<Korisnik>> {
        // handle login
        val result = dataSource.login(username, password)
        if (result.value is Result.Success<*>) {
            setLoggedInUser((result.value as Result.Success<*>).getData())
        }
        return result
    }

    companion object {
        @Volatile
        private var instance: LoginRepository? = null
        fun getInstance(dataSource: LoginDataSource): LoginRepository? {
            if (instance == null) {
                instance = LoginRepository(dataSource)
            }
            return instance
        }
    }
}